### PR TITLE
Fix card cap and sidebar mostly in dashboard design

### LINKED ISSUE #1055: Dashboard design interactions with cards
Bootstrap recommends changing card colors using `text-bg-{theme}` classes, but this approach reveals a number of tricky problems we've introduced with the dashboard design, as well as issues with the sidebar defaults. ([App on shinylive](https://shinylive.io/r/editor/#code=NobwRAdghgtgpmAXGKAHVA6ASmANGAYwHsIAXOMpMAGwEsAjAJykYE8AKAZwAtaJWAlAB0IdJiw71OY4aIbM27AObUArnFkjOcUhm1wAJuwAsAJgAEAanPHNETqSjlO5gDwBacwXZCw9KNq+uOa+BnAAZlCq1KRBIWCojLQwEnG+2sQQBql48XzhRGlgnKoEBHCcnEUA7iwQfEpF2RBKcIxNLADWRXRK3LFgdqq0bp6oUK0A+uG01NRQ9NRwPhDm5gDE5gCi0Itw5nQQ+-Rw1ETV5qRE5vrmaKjm1bSk3ESqpJfc+9k89EQsBnMYU4tCUEBEa02Lzg8HMAF5zFJJtD4OwAKzBdK8fhFKRidw-bh-AEEiqg8FgeHmABiAEEADIAZS2AlwEIOUFYb1Ik2IahgEEm1WYqBWazWTwMLyppgADLK2atxQBCVWoVSMTWIRApUUOJwVYLhVQQAikWgkLiOcgCcwgdnirwAsWO8UEeaVKkqdQ+MDkAAepHc9CU7hA+vIAF9fKyHa6CADJl8oGFGMo1MtfABZVh2iNwSNOxgGGOxpWujlc96TEFhfxpuMV2twetU5v1l0Vx2S6UIgCM8sVXddZ0YMO1tFQJRg7D7wTn5gAzLJh2sy6uDkQxzAJ1PVDOFwuAGwrrunx3n8yR0+yOzaRgANzao3MxtN5stfHVpGC3O-wW0SoLQgW0QCvEQtGxVhaXQdhhipYYALaJ9GDbZC2gEMBIwAXSAA))

<details>
<summary>Example App</summary>

```r
library(shiny)
library(bslib)
library(glue)

set.seed(42 + 4)

states <- c("base", "default", "primary", "secondary", "info", "success", "warning", "danger", "dark", "light")

ui <- page_fillable(
  # Enable line below to see app without the dashboard design
  # theme = bs_theme(5, "shiny", "bslib-dashboard-design" = FALSE),
  layout_column_wrap(
    width = 200,
    !!!purrr::map(states, function(state) {
      card(
        class = glue("text-bg-{state}"),
        card_header(glue("My {state} card")),
        layout_sidebar(
          sidebar = sidebar(
            width = 100,
            lorem::ipsum(1, 1, 3)
          ),
          lorem::ipsum(1, 1, 6)
        )
      )
    })
  )
)

server <- function(input, output, session) { }

shinyApp(ui = ui, server = server)
```

</details>

## With dashboard design (bslib default)

With the dashboard design, using `text-bg-{state}` results in the following card styles. In particular, the card cap is forced to have a white background and the sidebars are always `$body-color` text on a `$body-bg` background.

image

## With dashboard design disabled

With the dashboard design disabled, cards are close to the desired state, but the sidebar foreground color is forced to `$body-color` rather than inheriting.

image

## Base Bootstrap

With `bs_theme(5, "bootstrap")`, we see the following styles where again the sidebar foreground color is problematic.

image

### PR DESCRIPTION
**Pull request by gadenbuie**
Fixes #1055 and several issues with cards and sidebars, primarily in the Shiny preset but also in standard Bootstrap.

## After (dashboard design, light)

image

## After (dashboard design, dark)

image

## After (bslib, dashboard design disabled)

image

## Standard Bootstrap

image

## On mobile (bslib, shiny preset, dashboard design)

https://github.com/rstudio/bslib/assets/5420529/7b9478b4-3549-4261-8caf-8f539ffd1f69

### CODE REVIEW

File: inst/builtin/bs5/shiny/_rules.scss [hunk 1]
```diff
@@ -134,18 +134,16 @@ $bslib-checkbox-radio-margin-right: 0.35em !default;
     --bslib-dashboard-sidebar-main-bg: var(--#{$prefix}body-bg);
     --bslib-dashboard-main-bg: rgb(247, 247, 247);
     --#{$prefix}card-border-color: var(--bslib-dashboard-border-color-translucent, var(--#{$prefix}border-color-translucent));
-    --bslib-dashboard-card-cap-bg: var(--#{$prefix}body-bg);
     --bslib-dashboard-card-header-font-weight: #{$font-weight-semibold};
-    --bslib-sidebar-bg: var(--#{$prefix}body-bg);
-    --bslib-sidebar-fg: var(--#{$prefix}color);
+    --bslib-sidebar-bg: RGBA(var(--#{$prefix}body-bg-rgb), 0.1);
```
Comment (gadenbuie) [inline inst/builtin/bs5/shiny/_rules.scss:138]: Previously we were setting the sidebar background to the body background, but that's an absolute color. 

Now we use the body background color to lighten or darken the sidebar. This still gives us the white sidebar we want for the dashboard design without making the sidebar _always the body bg color_.

File: inst/builtin/bs5/shiny/_rules.scss [hunk 2]
```diff
@@ -134,18 +134,16 @@ $bslib-checkbox-radio-margin-right: 0.35em !default;
...
   }
 
   @include color-mode(dark) {
     --bslib-dashboard-main-bg: rgb(20, 20, 24);
   }
 
   .bslib-card {
-    --#{$prefix}card-cap-bg: var(--bslib-dashboard-card-cap-bg);
+    --#{$prefix}card-cap-bg: ;
```
Comment (gadenbuie) [inline inst/builtin/bs5/shiny/_rules.scss:146]: Bootstrap uses `$card-cap-bg: rgba(var(--#{$prefix}body-color-rgb), .03) !default;` and we don't want to get in the way of that -- we still want that value if dashboard design is disabled. This takes the `background-color: var(--bs-card-cap-bg)` out of play, letting the card cap inherit as needed, but also letting users still set `--bs-card-cap-bg` if they want.

File: inst/components/scss/sidebar.scss [hunk 3]
```diff
@@ -1,6 +1,6 @@
 // User-facing variables, use for theming
 $bslib-sidebar-bg: rgba(var(--bs-emphasis-color-rgb, 0,0,0), 0.05) !default;
-$bslib-sidebar-fg: var(--bs-emphasis-color, black) !default;
+$bslib-sidebar-fg: inherit !default;
```
Comment (gadenbuie) [inline inst/components/scss/sidebar.scss:3]: Let's just let the sidebar foreground color cascade

File: inst/components/scss/sidebar.scss [hunk 4]
```diff
@@ -368,6 +368,15 @@ $bslib-sidebar-column-sidebar: Min(calc(100% - var(--_padding-icon)), var(--_sid
...
       }
+
+      // Soften main area when sidebar is expanded over content
+      > .main {
+        opacity: 0.33;
+        transition: opacity var(--_transition-easing-x) var(--_transition-duration);
+      }
+      &.sidebar-collapsed > .main {
+        opacity: 1;
+      }
```
Comment (gadenbuie) [inline inst/components/scss/sidebar.scss:379]: Sidebars are now transparent in the dashboard design (as it already was in all others) and there's a nice glass effect. I just felt like we needed to dial down the main content area when the sidebar is expanded.

File: inst/builtin/bs5/shiny/_rules.scss (hunk 2 above)
Comment (cpsievert) [reply inst/builtin/bs5/shiny/_rules.scss:146]: This is very useful context that seems useful for development -- let's put it in a comment?

**Review by cpsievert — APPROVED** [review]
LGTM once minor comment is addressed, thanks!

**Pull request merged**